### DISCUSSION TITLE
Why so many context providers in a menu component?

### DISCUSSION THREAD

**Discussion by gnapse** [ariakit/ariakit#4067] [Questions] [answered]
Take a look at this screen recording:

https://github.com/user-attachments/assets/357b729c-740b-4f89-886d-78d1e8bc8dce

_(most of those with minified names are, in my experience trying this in a website with a development build, also several `Context.Provider` elements)_

I wonder why a menu needs that many context providers?

Inspecting the `value` props of many of these context providers, but they all seem store a very similar data structure. I cannot be sure if they're referentially the same value. But it looks like it could be.

https://github.com/user-attachments/assets/56785ee5-e1a2-425c-bc23-165943e12ecf

**Comment by diegohaz** [comment, marked as answer]
Component providers also provide contexts from their underlying components. For example, Menu provides contexts for Composite, Hovercard, Dialog, and Popover.

This might change in the future, not due to the number of context providers in the React tree (which has virtually no impact), but because of issues like #3754.

**Reply by gnapse** [reply]
Thanks for the info.

We're not too worried about an actual impact, as we only noticed it by using the component tree dev tool, and not because of any performance issue. However, it still looks like a too large amount of context providers, even for what you say. I'd expect a handful or two, say 10, 12, etc. But this is more around 100+, just for a single menu. It feels like it should not need that much, even after what you're saying.
